### PR TITLE
Implement hasPrefix and hasSuffix

### PR DESCRIPTION
**Pull request by glessard**
Using the _compareDeterministicUnicodeCollation method, also used by the == operator.

### CODE REVIEW

**Comment by gribozavr** [comment]
I think this might be close enough, although inefficient (traverses both strings twice).

This change requires tests.  We have tests for the Objective-C side in `test/1_stdlib/NSStringAPI.swift` and `validation-test/stdlib/String.swift`, please port those to start.

It would make sense to put the tests into `validation-test/stdlib/String.swift`, using `#if _runtime(ObjC)` to separate Objective-C dependent parts and un-XFAIL the test on Linux.

**Comment by glessard** [comment]
> On 11 déc. 2015, at 16:30, Dmitri Gribenko <notifications@github.com> wrote:
> 
> I think this might be close enough, although inefficient (traverses both strings twice).

There’s no claim of efficiency implied, but it could be worse; I’m fairly sure no new memory is allocated.

> This change requires tests. We have tests for the Objective-C side in test/1_stdlib/NSStringAPI.swift and validation-test/stdlib/String.swift, please port those to start.
> 
> It would make sense to put the tests into validation-test/stdlib/String.swift, using #if _runtime(ObjC) to separate Objective-C dependent parts and un-XFAIL the test on Linux.

Thanks for pointing me to those.

Guillaume Lessard

**Comment by parkera** [comment]
We should be careful with "close enough" implementations because the expectation is that it will behave the same as the Darwin version.

**Comment by parkera** [comment]
We have Foundation available on all platforms now. Can we make this use that implementation on all platforms as well?

**Comment by gribozavr** [comment]
@parkera That will create interesting circular dependencies in the build process.

**Comment by parkera** [comment]
If the build process is deficient then we should fix that instead of forcing an incompatibility onto clients of the API.

**Comment by parkera** [comment]
Or we do the obvious thing and put the implementation in an extension on String in Foundation.

File: stdlib/public/core/StringLegacy.swift [hunk 1]
```diff
@@ -95,8 +95,33 @@ extension String {
   }
 }
 #else
-// FIXME: Implement hasPrefix and hasSuffix without objc
-// rdar://problem/18878343
+extension String {
+  /// Returns `true` iff `self` begins with `prefix`.
```
Comment (nickskull) [inline stdlib/public/core/StringLegacy.swift]: This should probably be just "if" and the same thing below :)

Comment (glessard) [inline stdlib/public/core/StringLegacy.swift]: "iff" (if-and-only-if) is a stronger statement than "if". It is correct here.

Comment (nickskull) [inline stdlib/public/core/StringLegacy.swift]: I stand corrected. My apologies.

File: stdlib/public/core/StringLegacy.swift [hunk 2]
```diff
@@ -95,8 +95,33 @@ extension String {
...
+  public func hasPrefix(prefix: String) -> Bool {
+    if prefix.isEmpty { return false }
```
Comment (ZuymantO) [inline stdlib/public/core/StringLegacy.swift]: Theoretically *str represent str or <SOME STRING>str that it why I think this should return true. 
Example
--
Given a string s and a literal string "Hello", if I want to check *with hasPrefix*, that adding "Hello" with doesn't changes anything and remains "Hello". 
So, let concStr = "\(s)Hello"; [...] 
conStr.hasPrefix("") shouldn't returns true ?

Comment (glessard) [inline stdlib/public/core/StringLegacy.swift]: The goal here is to reproduce the NSString functionality. If you enter `"String".hasPrefix("")` in an Xcode Swift playground (where you're getting the Objective-C implementation at this time), you'll see that the result is `false`. `"String".hasPrefix("S")` is `true`, as expected.

In any event, the empty String as a parameter to `hasPrefix` is what is generally known as a trivial solution. Technically it is a substring of the operand, but that's not useful information.

**Comment by glessard** [comment]
@gribozavr The relevant test is ported from test/1_stdlib/NSStringAPI.swift.

Neither the original test nor the ported test are particularly satisfying, since ultimately the operation is tested with a different version of itself.

**Comment by glessard** [comment]
@parkera From what I ran into while doing this, I'd advocate to eliminate any way to reach Foundation-based comparisons from Swift.String. If a variable is being used as NSString or CFString, it's okay to use them, but otherwise there is at least one corner case where the Foundation-based comparison is not consistent with the Swift.String model. See https://bugs.swift.org/browse/SR-243

**Comment by parkera** [comment]
If we want to discuss dropping the NSString methods from String, that would be something to discuss on the evolution list. In the meantime, my primary concern is making sure developers don't get different answers for the same methods on Linux vs OS X.

**Comment by dabrahams** [comment]
> On Dec 16, 2015, at 10:13 AM, Tony Parker <notifications@github.com> wrote:
> 
> If we want to discuss dropping the NSString methods from String, that would be something to discuss on the evolution list. In the meantime, my primary concern is making sure developers don't get different answers for the same methods on Linux vs OS X.
> 

+1

-Dave

**Comment by glessard** [comment]
After porting the corrected test from https://github.com/apple/swift/pull/629, the Foundation-based calls no longer pass, but the ICU-based calls do. Not sure where to go from here.

**Comment by glessard** [comment]
Would you like a consolidated PR with this?

**Comment by kballard** [comment]
Incidentally, it turns out that `_compareDeterministicUnicodeCollation` actually bridges into ObjC if `_runtime(_ObjC)`. I don't know why it does this at all, unless `_compareDeterministicUnicodeCollation` doesn't actually match `NSString` behavior (but you'd think it would). This means that anything that uses `_compareDeterministicUnicodeCollation` ends up actually allocating a new `_NSContiguousString` if the string in question isn't already backed by a Cocoa string.

**Comment by dabrahams** [comment]
> On Jan 3, 2016, at 4:00 PM, Kevin Ballard <notifications@github.com> wrote:
> 
> Incidentally, it turns out that _compareDeterministicUnicodeCollation actually bridges into ObjC if _runtime(_ObjC). I don't know why it does this at all, unless _compareDeterministicUnicodeCollation doesn't actually match NSString behavior (but you'd think it would). This means that anything that uses _compareDeterministicUnicodeCollation ends up actually allocating a new _NSContiguousString if the string in question isn't already backed by a Cocoa string.
> 
Except maybe on Linux, every string is backed by a Cocoa string.  The Swift-native String backing buffer is-a NSString.

-Dave

**Comment by kballard** [comment]
@dabrahams I don't see how that's possible. The Swift-native String buffer is a `_StringBuffer`, which uses a `_HeapBuffer<_StringBufferIVars, UTF16.CodeUnit>` as its storage, which in turn uses `_HeapBufferStorage<_StringBufferIVars, UTF16.CodeUnit>`, which is a class that inherits from `NonObjectiveCBase`. It doesn't seem possible for this to be layout-compatible with an `NSString`, because Swift classes always have the type info and retain counts at the beginning of their allocated storage (which is precisely where the isa pointer needs to go for Obj-C). Plus of course `_StringBufferIVars` doesn't even attempt to include an `isa` pointer.

You can also trivially demonstrate that a Swift string is not Cocoa-compatible with the following code:

```swift
var s = "test"
s += "!" // force it to copy to mutable memory
print(s._core.nativeBuffer)
print(s._core.cocoaBuffer)
```

If it was a Cocoa string, it would print `nil` and then the Cocoa string. But it's not, it's a native string, so it prints the `_StringBuffer` and then it prints `nil`. And since `_stdlib_binary_bridgeToObjectiveCImpl` only avoids creating a new `_NSContiguousString` if `_core.cocoaBuffer` is non-`nil` (and `_swift_stdlib_CFStringGetLength` reports the same length as `_core.count`, not entirely sure why), this should demonstrate that Swift-native strings always end up creating a new `_NSContiguousString` whenever they bridge into Obj-C.

So this all comes back to the question of why String bridges into Obj-C when it has a perfectly functional non-ObjC implementation available. It would make sense to bridge if it already has a Cocoa buffer, but it doesn't test that, it merely tests `#if _runtime(_ObjC)`.

Incidentally, there's a bit of even more weirdness here. When comparing two strings with `_runtime(_ObjC)`, if both strings are ASCII, it falls back to a separate non-ObjC comparison routine that's basically a wrapper around `memcmp`. The documentation says that Unicode actually defines the sort order differently for some punctuation than the ASCII value order, but that Foundation uses the ASCII value order. This is strange because it suggests that two Swift strings can actually compare differently depending on whether the ObjC runtime is present (e.g. `"$" < "*"` prints `true` with the Obj-C runtime but it should print `false` with the root collation order from Unicode). What's even odder about this is the Unicode Collation Algorithm can actually be tailored, and in fact the `en_US_POSIX` locale (also known as the `C` locale) already tailors the collation specifically to make ASCII characters sort in ASCII value order. So I'm surprised we don't just use an ICU collator with the collation table tailored to make ASCII characters sort in ASCII order the same way `en_US_POSIX` does (I assume `en_US_POSIX` inherits other collation tweaks from `en_US` or else I'd suggest just using `en_US_POSIX` directly, although I don't understand ICU well enough to know if that's actually true).

**Comment by glessard** [comment]
As far as I can tell, the OSX swift build doesn't link ICU at all. I attempted earlier to bypass the call to `_stdlib_compareNSStringDeterministicUnicodeCollation` within `_compareDeterministicUnicodeCollation` (by inserting `#if false` at line 371 of String.swift); that resulted in a link error.

**Comment by adrfer** [comment]
@glessard, really nice implementation! 👌

I was just wondering which one is more Swift-y at the `hasPrefix` function?

```swift
if prefix.isEmpty { return false }
```

or

```swift
guard !prefix.isEmpty else { return false }
```

Any thoughts?

Same goes for `hasSuffix`.

**Comment by glessard** [comment]
@adrfer I'm sure it could be argued both ways, but I think it is a special condition rather than a precondition. If this eventually called into `CFStringFind`, the `if` statement could conceivably be removed; this eventually calls into `CFStringCompare`, and two empty strings would compare as equal, which is not the desired answer. (Plus, `guard` leads to a double-negative.)

**Comment by adrfer** [comment]
@glessard indeed! I'm just trying to see what's more idiomatic. In this case `guard`'s double negative isn't really cool, and its version is even longer than the `if` version. Anyways, Swift is evolving fast and looking forward to seeing what preferred style the community will come up with.

**Comment by glessard** [comment]
Rebased and re-submitted as https://github.com/apple/swift/pull/1031

**Comment by kballard** [comment]
You know, you can just force-push to your branch and GitHub will update the PR. That way all the previous discussion will be preserved.

**Comment by glessard** [comment]
@kballard I didn't know that could work. It doesn't look like I can reopen, though.

**Comment by glessard** [comment]
Rebased here. Thanks for letting me know this is possible, @kballard

**Comment by CodaFi** [comment]
@glessard Given the current implementation, do you think this pull request is still necessary/relevant?

**Comment by glessard** [comment]
@CodaFi Good question. I was considering re-implementing the tests part of this based on the newer work; it would at least illuminate some differences between an ICU-based implementation and the CFString-based one. I think the discussion is still of interest for the (hopefully) forthcoming String rethink.

**Comment by CodaFi** [comment]
Sounds good; we'll always take more tests.  To preserve the discussion here, I'm going to close this pull request.  Please open another if you decide to cherry pick the tests.

**Pull request closed without merging**